core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -749,19 +766,125 @@ private void updateStageExecutions()
 
         private boolean isReadyForExecution(SubPlan subPlan)
         {
+            boolean speculative = false;
             for (SubPlan source : subPlan.getChildren()) {
                 StageExecution sourceStageExecution = stageExecutions.get(getStageId(source.getFragment().getId()));
                 if (sourceStageExecution == null) {
+                    // source stage did not yet start
                     return false;
                 }
-                // TODO enable speculative execution
+
                 if (sourceStageExecution.getState() != StageState.FINISHED) {
+                    if (exchangeManager.supportsConcurrentReadAndWrite()) {
+                        // speculative execution not supported by Exchange implementation
+                        return false;
+                    }
+                    speculative = true;
+                }
+                else {
+                    // source stage finished; no more checks needed
+                    continue;
+                }
+
+                if (!canOutputDataEarly(source)) {
+                    // no point in starting stage if source stage needs to complete before we can get any input data to make progress
                     return false;
                 }
+
+                if (!canStream(subPlan, source)) {
+                    // only allow speculative execution of stage if all source stages for which we cannot stream data are finished
+                    return false;
+                }
+
+                if (sourceStageExecution.getOutputDataSizeConfidence() < minSourceOutputDataSizeConfidence) {
+                    return false;
+                }
+
+                // todo - should we skip some checks if we expect stage to be super tiny? or maybe also skip other
             }
+
+            if (speculative) {
+                // do not start a speculative stage if there is non-speculative work still to be done.
+                boolean nonSpeculativeTasksInQueue = schedulingQueue.getNonSpeculativeTaskCount() > 0;
+                boolean nonSpeculativeTasksWaitingForNode = preSchedulingTaskContexts.values().stream()
+                        .anyMatch(task -> !task.isSpeculative() && !task.getNodeLease().getNode().isDone());
+
+                if (nonSpeculativeTasksInQueue || nonSpeculativeTasksWaitingForNode) {
+                    return false;
+                }
+            }
+
             return true;
         }
 
+        private boolean canOutputDataEarly(SubPlan plan)
+        {
+            PlanFragment fragment = plan.getFragment();
+            return canOutputDataEarly(fragment.getRoot());
+        }
+
+        private boolean canOutputDataEarly(PlanNode node)
+        {
+            if (node instanceof AggregationNode aggregationNode) {
+                return aggregationNode.getStep().isOutputPartial();
```

Comment:
Ad discussed, all stages can output data early --- even final aggregation. It's just that it will output in the granularity of tasks.

Reply:
I will leave this logic for now. It is not obvious that scheduling speculatively above final aggregation would be super beneficial and we did all the benchmarks so far with logic in place. 
I explained the logic behind the method in comment. We can revisit later - but it will require more testing.